File: src/main/kotlin/xyz/mahmoudahmed/dsl/NestingTargetBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.function.BiFunction;

/**
 * Builder for specifying the target of nesting operations.
 */
public class NestingTargetBuilder {
    private final TargetBuilder parent;
    private final String sourcePattern;
    private final String structureName;
    private final String fieldNameExtractor;
    private final BiFunction<String, Object, Object> valueTransformation;
    private final boolean isCollection;

    NestingTargetBuilder(TargetBuilder parent, String sourcePattern, String structureName,
                         String fieldNameExtractor, BiFunction<String, Object, Object> valueTransformation,
                         boolean isCollection) {
        this.parent = parent;
        this.sourcePattern = sourcePattern;
        this.structureName = structureName;
        this.fieldNameExtractor = fieldNameExtractor;
        this.valueTransformation = valueTransformation;
        this.isCollection = isCollection;
    }

    /**
     * Specify the target path for the nested structure.
     */
    public TargetBuilder to(String targetPath) {
        MappingRule rule = new NestingMappingRule(sourcePattern, targetPath, structureName,
                fieldNameExtractor, valueTransformation, isCollection);
        parent.addRule(rule);
        return parent;
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/structure/NestingTargetBuilder.kt
package xyz.mahmoudahmed.dsl.structure

import xyz.mahmoudahmed.dsl.builders.TargetBuilder

/**
 * Builder for specifying the target of nesting operations.
 */
class NestingTargetBuilder(
    private val parent: TargetBuilder,
    private val sourcePattern: String,
    private val structureName: String,
    private val fieldNameExtractor: String?,
    private val valueTransformation: ((String, Any) -> Any)?,
    private val isCollection: Boolean,
    private val includeNullValues: Boolean,
    private val strictTypeHandling: Boolean
) {
    /**
     * Specify the target path for the nested structure.
     *
     * @param targetPath Path where the nested structure will be created
     * @return The parent target builder
     */
    fun to(targetPath: String): TargetBuilder {
        validateTargetPath(targetPath)
        val rule = NestingMappingRule(
            sourcePattern, targetPath, structureName,
            fieldNameExtractor, valueTransformation, isCollection,
            includeNullValues, strictTypeHandling
        )
        println()
        parent.addRule(rule)
        return parent
    }

    private fun validateTargetPath(path: String) {
        if (path.contains("*")) {
            throw IllegalArgumentException("Target path cannot contain wildcards: $path")
        }
    }
}